File: 77주차/1012.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const [dx, dy] = [
  [-1, 1, 0, 0],
  [0, 0, -1, 1],
];
const bfs = (x, y) => {
  visited[x][y] = 1;
  const q = [];
  q.push([x, y]);
  while (q.length > 0) {
    const [x, y] = q.shift();

    for (let i = 0; i < 4; i++) {
      const [nx, ny] = [x + dx[i], y + dy[i]];
      if (0 <= nx && nx < n && 0 <= ny && ny < m) {
        if (!visited[nx][ny] && g[nx][ny] == 1) {
          visited[nx][ny] = 1;
          q.push([nx, ny]);
        }
      }
    }
  }
};

const solution = () => {
  let answer = 0;
  for (let x = 0; x < n; x++) {
    for (let y = 0; y < m; y++) {
      if (!visited[x][y] && g[x][y] == 1) {
        bfs(x, y);
        answer += 1;
      }
    }
  }
  return answer;
};
let [m, n, k] = [null, null, null];
let tc = null;
let visited = [];
let g = [];
let tcCnt = 0;
let cnt = 0;
rl.on("line", (line) => {
  if (!tc) {
    tc = +line;
  } else if (!n) {
    [m, n, k] = line.split(" ").map((e) => +e);
    g = Array(n)
      .fill()
      .map((e) => Array(m).fill(0));
    visited = Array(n)
      .fill()
      .map((e) => Array(m).fill(0));
  } else if (cnt < k) {
    const [y, x] = line.split(" ").map((e) => +e); // 가로 세로 위치 순으로 입력 받음. 주의
    g[x][y] = 1;
    cnt += 1;

    if (cnt == k) {
      console.log(solution());
      [m, n, k] = [null, null, null];
      cnt = 0;
      tcCnt += 1;
      if (tc == tcCnt) rl.close();
    }
  }
});
rl.on("close", () => {
  process.exit();
});

File: 80주차/1012.py
import sys
sys.setrecursionlimit(10*1000)# 파이썬은 기본 재귀 제한이 1000. 이 문제는 2500번까지 가능.
dx = [-1,1,0,0]
dy = [0, 0,-1,1]
def dfs(x,y):
    visited[x][y] = 1
    for i in range(4):
        nx,ny = x+dx[i],y+dy[i]
        if 0<= nx < n and 0 <= ny < m and not visited[nx][ny] and g[nx][ny] == 1:
            dfs(nx,ny)

T = int(input())
for _ in range(T):
    m,n,k = map(int,input().split())
    g = [[0]*m for _ in range(n)]
    for _ in range(k):
        y,x = map(int,input().split())
        g[x][y] = 1
    cnt = 0
    visited = [[0]*m for _ in range(n)]
    for x in range(n):
        for y in range(m):
            if not visited[x][y] and g[x][y] == 1:
                dfs(x,y)
                cnt += 1# 덩어리 개수 세기
    print(cnt)
# 백준 유기농배추.
